### DISCUSSION TITLE
Simulation Study - NB + TFIDF ngram_max=(1 to 4)

### DISCUSSION THREAD

**Discussion by rohitgarud** [asreview/asreview#1339] [Show and tell]
This study was performed to see if bigrams or trigrams perform better than unigrams in TFIDF features. 
**Classifier model**: Naive Bayes with default parameters
**Query strategy**: Max 
**Balance strategy**: Double 
**Feature Extraction**: TFIDF with ngram_max values ranging from 1 to 4

**Conclusion**: Unigrams outperformed most of the time and bigrams performed well on a few of the datasets. However, the performance increase by bigrams was not sufficient enough to use them as default. Improvement with n-grams higher than 1 came at a cost of increased time due to a significantly larger vocabulary than unigrams. So unigrams are a good choice to be used with NB. This means the feature independence assumption of NB holds beautifully for unigrams in most cases.

Simulations were performed using [Asreview-makita extension](https://github.com/asreview/asreview-makita) with a custom template and script.

**Comment by rohitgarud** [comment]
| Dataset                 | TFIDF ngram_max | recall_0.1  | recall_0.25 | recall_0.5  | recall_0.75 | recall_0.9  | wss_0.95     | erf_0.1      | atd         | td_last |
|-------------------------|-----------------|-------------|-------------|-------------|-------------|-------------|--------------|--------------|-------------|---------|
| ACEInhibitors           | 1               | 0.85        | 0.95        | 0.95        | 0.975       | 1           | 0.781667978  | 0.75         | 190.325     | 2031    |
|                         | 2               | 0.825       | 0.95        | 0.95        | 0.975       | 1           | 0.773013375  | 0.725        | 195.875     | 2154    |
|                         | 3               | 0.825       | 0.95        | 0.95        | 0.95        | 1           | 0.785601888  | 0.725        | 194.95      | 2187    |
|                         | 4               | 0.8         | 0.95        | 0.95        | 0.95        | 1           | 0.791109363  | 0.7          | 201.45      | 2189    |
| ADHD                    | 1               | 0.842105263 | 0.842105263 | 0.947368421 | 1           | 1           | 0.446407538  | 0.736842105  | 96.78947368 | 572     |
|                         | 2               | 0.789473684 | 0.842105263 | 0.947368421 | 0.947368421 | 1           | 0.49352179   | 0.684210526  | 100.7368421 | 688     |
|                         | 3               | 0.789473684 | 0.842105263 | 0.947368421 | 0.947368421 | 1           | 0.500588928  | 0.684210526  | 97.84210526 | 639     |
|                         | 4               | 0.789473684 | 0.842105263 | 0.947368421 | 0.947368421 | 1           | 0.524146054  | 0.684210526  | 95.42105263 | 645     |
| Antihistamines          | 1               | 0.333333333 | 0.733333333 | 0.866666667 | 0.866666667 | 0.933333333 | 0.090909091  | 0.266666667  | 72.33333333 | 284     |
|                         | 2               | 0.333333333 | 0.666666667 | 0.866666667 | 0.866666667 | 0.933333333 | 0.11038961   | 0.266666667  | 78.6        | 303     |
|                         | 3               | 0.333333333 | 0.666666667 | 0.866666667 | 0.866666667 | 0.933333333 | 0.133116883  | 0.266666667  | 78.6        | 303     |
|                         | 4               | 0.333333333 | 0.666666667 | 0.866666667 | 0.866666667 | 0.933333333 | 0.133116883  | 0.266666667  | 82.6        | 303     |
| Appenzeller-Herzog_2020 | 1               | 0.821428571 | 0.928571429 | 0.964285714 | 1           | 1           | 0.722978847  | 0.714285714  | 263.4642857 | 1762    |
|                         | 2               | 0.857142857 | 0.892857143 | 0.964285714 | 1           | 1           | 0.592871631  | 0.75         | 261.1071429 | 1871    |
|                         | 3               | 0.857142857 | 0.892857143 | 0.964285714 | 1           | 1           | 0.56563315   | 0.75         | 276         | 2014    |
|                         | 4               | 0.857142857 | 0.892857143 | 0.964285714 | 1           | 1           | 0.549116198  | 0.75         | 282.9642857 | 2074    |
| AtypicalAntipsychotics  | 1               | 0.351724138 | 0.655172414 | 0.868965517 | 0.937931034 | 0.979310345 | 0.168157424  | 0.255172414  | 275.1241379 | 1106    |
|                         | 2               | 0.35862069  | 0.648275862 | 0.868965517 | 0.931034483 | 0.972413793 | 0.15921288   | 0.262068966  | 272.4896552 | 1113    |
|                         | 3               | 0.35862069  | 0.64137931  | 0.868965517 | 0.937931034 | 0.972413793 | 0.186046512  | 0.262068966  | 273.2758621 | 1100    |
|                         | 4               | 0.344827586 | 0.634482759 | 0.875862069 | 0.910344828 | 0.972413793 | 0.178890877  | 0.248275862  | 276.5241379 | 1093    |
| Bannach-Brown_2019      | 1               | 0.523297491 | 0.856630824 | 0.931899642 | 0.971326165 | 0.989247312 | 0.403817177  | 0.422939068  | 293.0215054 | 1940    |
|                         | 2               | 0.534050179 | 0.856630824 | 0.9390681   | 0.964157706 | 0.982078853 | 0.398292315  | 0.433691756  | 299.9820789 | 1921    |
|                         | 3               | 0.551971326 | 0.849462366 | 0.94265233  | 0.964157706 | 0.985663082 | 0.350577599  | 0.451612903  | 297.311828  | 1903    |
|                         | 4               | 0.551971326 | 0.856630824 | 0.94265233  | 0.971326165 | 0.985663082 | 0.342039176  | 0.451612903  | 295.4444444 | 1920    |
| BetaBlockers            | 1               | 0.707317073 | 0.926829268 | 0.926829268 | 0.951219512 | 0.975609756 | 0.669082126  | 0.609756098  | 241.8780488 | 1956    |
|                         | 2               | 0.682926829 | 0.902439024 | 0.951219512 | 0.951219512 | 1           | 0.657004831  | 0.585365854  | 243.5609756 | 1823    |
|                         | 3               | 0.707317073 | 0.804878049 | 0.951219512 | 0.951219512 | 0.975609756 | 0.636714976  | 0.609756098  | 251.1219512 | 1884    |
|                         | 4               | 0.731707317 | 0.804878049 | 0.951219512 | 0.951219512 | 1           | 0.623188406  | 0.634146341  | 256.1463415 | 1853    |
| Bos_2018                | 1               | 1           | 1           | 1           | 1           | 1           | 0.833739554  | 0.9          | 76          | 273     |
|                         | 2               | 1           | 1           | 1           | 1           | 1           | 0.832346797  | 0.9          | 78.3        | 348     |
|                         | 3               | 1           | 1           | 1           | 1           | 1           | 0.831824513  | 0.9          | 90.4        | 489     |
|                         | 4               | 1           | 1           | 1           | 1           | 1           | 0.830431755  | 0.9          | 104.1       | 527     |
| CalciumChannelBlockers  | 1               | 0.272727273 | 0.686868687 | 0.898989899 | 0.95959596  | 0.95959596  | 0.228618421  | 0.171717172  | 268.3636364 | 1176    |
|                         | 2               | 0.282828283 | 0.696969697 | 0.909090909 | 0.95959596  | 0.95959596  | 0.214638158  | 0.181818182  | 280.9292929 | 1202    |
|                         | 3               | 0.272727273 | 0.747474747 | 0.929292929 | 0.95959596  | 0.95959596  | 0.209703947  | 0.171717172  | 279.2525253 | 1199    |
|                         | 4               | 0.282828283 | 0.727272727 | 0.929292929 | 0.95959596  | 0.95959596  | 0.206414474  | 0.181818182  | 276.9191919 | 1204    |
| Estrogens               | 1               | 0.278481013 | 0.620253165 | 0.911392405 | 0.962025316 | 0.962025316 | 0.24863388   | 0.17721519   | 89.41772152 | 351     |
|                         | 2               | 0.278481013 | 0.632911392 | 0.886075949 | 0.962025316 | 0.962025316 | 0.289617486  | 0.17721519   | 89.75949367 | 357     |
|                         | 3               | 0.291139241 | 0.620253165 | 0.886075949 | 0.962025316 | 0.962025316 | 0.281420765  | 0.189873418  | 87.15189873 | 354     |
|                         | 4               | 0.303797468 | 0.594936709 | 0.886075949 | 0.962025316 | 0.962025316 | 0.295081967  | 0.202531646  | 92.21518987 | 356     |
| Hall_2012               | 1               | 0.990291262 | 1           | 1           | 1           | 1           | 0.910764396  | 0.893203883  | 134.2718447 | 1767    |
|                         | 2               | 0.990291262 | 1           | 1           | 1           | 1           | 0.912560332  | 0.893203883  | 122.0485437 | 1660    |
|                         | 3               | 0.990291262 | 1           | 1           | 1           | 1           | 0.913121562  | 0.893203883  | 124.8058252 | 1655    |
|                         | 4               | 0.990291262 | 1           | 1           | 1           | 1           | 0.91233584   | 0.893203883  | 126.3495146 | 1665    |
| Kitchenham_2010         | 1               | 0.659090909 | 0.863636364 | 0.977272727 | 1           | 1           | 0.652173913  | 0.568181818  | 200.1363636 | 983     |
|                         | 2               | 0.590909091 | 0.886363636 | 0.977272727 | 1           | 1           | 0.616921269  | 0.5          | 218.4318182 | 928     |
|                         | 3               | 0.590909091 | 0.863636364 | 0.977272727 | 1           | 1           | 0.586956522  | 0.5          | 205.3863636 | 942     |
|                         | 4               | 0.590909091 | 0.886363636 | 0.977272727 | 1           | 1           | 0.599882491  | 0.5          | 203.6818182 | 939     |
| Kwok_2020               | 1               | 0.68907563  | 0.932773109 | 0.991596639 | 1           | 1           | 0.675272287  | 0.588235294  | 227.092437  | 1306    |
|                         | 2               | 0.731092437 | 0.949579832 | 0.991596639 | 1           | 1           | 0.696651876  | 0.630252101  | 212.0672269 | 1353    |
|                         | 3               | 0.739495798 | 0.932773109 | 0.991596639 | 1           | 1           | 0.669624849  | 0.638655462  | 213.8907563 | 1306    |
|                         | 4               | 0.722689076 | 0.932773109 | 0.991596639 | 1           | 1           | 0.674868899  | 0.621848739  | 217.3781513 | 1306    |
| Nagtegaal_2019          | 1               | 0.64        | 0.84        | 0.99        | 1           | 1           | 0.528507685  | 0.54         | 237.19      | 1088    |
|                         | 2               | 0.61        | 0.92        | 0.94        | 1           | 1           | 0.440257809  | 0.51         | 229.82      | 1173    |
|                         | 3               | 0.61        | 0.92        | 0.98        | 1           | 1           | 0.48636589   | 0.51         | 224.02      | 1159    |
|                         | 4               | 0.62        | 0.92        | 0.98        | 1           | 1           | 0.460089241  | 0.52         | 221.99      | 1159    |
| NSAIDS                  | 1               | 0.6         | 0.975       | 0.975       | 0.975       | 0.975       | 0.698209719  | 0.5          | 46.775      | 365     |
|                         | 2               | 0.5         | 0.925       | 0.975       | 0.975       | 0.975       | 0.685421995  | 0.4          | 49.95       | 370     |
|                         | 3               | 0.5         | 0.925       | 0.975       | 0.975       | 0.975       | 0.68286445   | 0.4          | 49.425      | 374     |
|                         | 4               | 0.5         | 0.925       | 0.975       | 0.975       | 0.975       | 0.68797954   | 0.4          | 50.125      | 366     |
| Opiods                  | 1               | 0           | 0.071428571 | 0.928571429 | 1           | 1           | 0.467328803  | -0.071428571 | 717.8571429 | 1290    |
|                         | 2               | 0           | 0.071428571 | 0.928571429 | 1           | 1           | 0.480397282  | -0.071428571 | 746         | 1202    |
|                         | 3               | 0           | 0.071428571 | 0.928571429 | 1           | 1           | 0.527966545  | -0.071428571 | 665.6428571 | 1168    |
|                         | 4               | 0           | 0.071428571 | 0.928571429 | 1           | 1           | 0.524307371  | -0.071428571 | 675.8571429 | 1206    |
| OralHypoglycemics       | 1               | 0.237037037 | 0.525925926 | 0.77037037  | 0.940740741 | 0.977777778 | 0.189620758  | 0.140740741  | 155.4444444 | 493     |
|                         | 2               | 0.266666667 | 0.555555556 | 0.748148148 | 0.903703704 | 0.977777778 | 0.169660679  | 0.17037037   | 155.7777778 | 498     |
|                         | 3               | 0.259259259 | 0.511111111 | 0.748148148 | 0.911111111 | 0.977777778 | 0.155688623  | 0.162962963  | 157.0444444 | 487     |
|                         | 4               | 0.259259259 | 0.518518519 | 0.748148148 | 0.903703704 | 0.977777778 | 0.151696607  | 0.162962963  | 157.4888889 | 485     |
| ProtonPumpInhibitors    | 1               | 0.66        | 0.8         | 0.94        | 0.98        | 1           | 0.480841473  | 0.56         | 191.96      | 1159    |
|                         | 2               | 0.68        | 0.8         | 0.94        | 0.98        | 1           | 0.47257701   | 0.58         | 193.46      | 1124    |
|                         | 3               | 0.72        | 0.8         | 0.92        | 0.98        | 1           | 0.429000751  | 0.62         | 193.76      | 1153    |
|                         | 4               | 0.74        | 0.8         | 0.92        | 0.98        | 1           | 0.401202104  | 0.64         | 192.06      | 1161    |
| Radjenovic_2013         | 1               | 0.957446809 | 1           | 1           | 1           | 1           | 0.84078026   | 0.85106383   | 189.5744681 | 1321    |
|                         | 2               | 0.957446809 | 1           | 1           | 1           | 1           | 0.851617206  | 0.85106383   | 191.7021277 | 1414    |
|                         | 3               | 0.957446809 | 1           | 1           | 1           | 1           | 0.86312104   | 0.85106383   | 183.9787234 | 1288    |
|                         | 4               | 0.957446809 | 1           | 1           | 1           | 1           | 0.860786929  | 0.85106383   | 180.3617021 | 1180    |
| SkeletalMuscleRelaxants | 1               | 0           | 0.25        | 0.625       | 0.75        | 1           | -0.014625229 | -0.125       | 782.625     | 1360    |
|                         | 2               | 0           | 0.125       | 0.625       | 0.75        | 1           | -0.02559415  | -0.125       | 814.875     | 1379    |
|                         | 3               | 0           | 0.125       | 0.5         | 0.75        | 1           | 0.035953687  | -0.125       | 834.25      | 1278    |
|                         | 4               | 0           | 0.125       | 0.5         | 0.75        | 1           | -0.048750762 | -0.125       | 866         | 1417    |
| Statins                 | 1               | 0.726190476 | 0.785714286 | 0.94047619  | 0.988095238 | 1           | 0.470690153  | 0.630952381  | 445.0357143 | 2921    |
|                         | 2               | 0.714285714 | 0.785714286 | 0.952380952 | 0.988095238 | 0.988095238 | 0.520646838  | 0.619047619  | 436.1785714 | 3210    |
|                         | 3               | 0.714285714 | 0.797619048 | 0.952380952 | 0.988095238 | 0.988095238 | 0.501588218  | 0.619047619  | 447.9404762 | 3188    |
|                         | 4               | 0.726190476 | 0.797619048 | 0.952380952 | 0.988095238 | 0.988095238 | 0.505342189  | 0.630952381  | 445.7857143 | 3209    |
| Triptans                | 1               | 0.565217391 | 0.739130435 | 0.956521739 | 1           | 1           | 0.520179372  | 0.47826087   | 98.56521739 | 349     |
|                         | 2               | 0.565217391 | 0.739130435 | 0.956521739 | 1           | 1           | 0.526158445  | 0.47826087   | 92.65217391 | 338     |
|                         | 3               | 0.608695652 | 0.739130435 | 0.956521739 | 1           | 1           | 0.520179372  | 0.52173913   | 92          | 345     |
|                         | 4               | 0.608695652 | 0.739130435 | 0.956521739 | 1           | 1           | 0.511210762  | 0.52173913   | 92.26086957 | 361     |
| UrinaryIncontinence     | 1               | 0.461538462 | 0.743589744 | 0.871794872 | 1           | 1           | 0.381538462  | 0.358974359  | 60.92307692 | 242     |
|                         | 2               | 0.487179487 | 0.769230769 | 0.871794872 | 1           | 1           | 0.415384615  | 0.384615385  | 58.82051282 | 231     |
|                         | 3               | 0.461538462 | 0.769230769 | 0.897435897 | 1           | 1           | 0.427692308  | 0.358974359  | 58.38461538 | 231     |
|                         | 4               | 0.487179487 | 0.743589744 | 0.871794872 | 1           | 1           | 0.424615385  | 0.384615385  | 57.38461538 | 206     |
| van_de_Schoot_2017      | 1               | 1           | 1           | 1           | 1           | 1           | 0.888475836  | 0.904761905  | 101.5714286 | 549     |
|                         | 2               | 1           | 1           | 1           | 1           | 1           | 0.890092129  | 0.904761905  | 94.26190476 | 285     |
|                         | 3               | 1           | 1           | 1           | 1           | 1           | 0.884111847  | 0.904761905  | 102.4285714 | 260     |
|                         | 4               | 1           | 1           | 1           | 1           | 1           | 0.882333926  | 0.904761905  | 108.452381  | 269     |
| van_Dis_2020            | 1               | 0.652777778 | 0.888888889 | 0.972222222 | 1           | 1           | 0.650442882  | 0.555555556  | 1191.263889 | 5620    |
|                         | 2               | 0.666666667 | 0.888888889 | 1           | 1           | 1           | 0.642498402  | 0.569444444  | 1081.069444 | 5072    |
|                         | 3               | 0.666666667 | 0.875       | 1           | 1           | 1           | 0.619852068  | 0.569444444  | 1096.444444 | 5384    |
|                         | 4               | 0.666666667 | 0.875       | 1           | 1           | 1           | 0.619395489  | 0.569444444  | 1105.236111 | 5336    |
| Wahono_2015             | 1               | 0.967213115 | 0.983606557 | 1           | 1           | 1           | 0.863428571  | 0.868852459  | 197.8032787 | 2098    |
|                         | 2               | 0.967213115 | 0.983606557 | 1           | 1           | 1           | 0.869        | 0.868852459  | 176.147541  | 2000    |
|                         | 3               | 0.967213115 | 0.983606557 | 1           | 1           | 1           | 0.872571429  | 0.868852459  | 178.5245902 | 1846    |
|                         | 4               | 0.967213115 | 0.983606557 | 1           | 1           | 1           | 0.870428571  | 0.868852459  | 186.9344262 | 1811    |
| Wolters_2018            | 1               | 0.888888889 | 1           | 1           | 1           | 1           | 0.794299382  | 0.777777778  | 162.2222222 | 997     |
|                         | 2               | 0.944444444 | 1           | 1           | 1           | 1           | 0.858281842  | 0.833333333  | 132.3333333 | 945     |
|                         | 3               | 0.944444444 | 1           | 1           | 1           | 1           | 0.874028304  | 0.833333333  | 134.8888889 | 957     |
|                         | 4               | 0.944444444 | 1           | 1           | 1           | 1           | 0.87024118   | 0.833333333  | 133.8888889 | 906     |

**Reply by J535D165** [reply]
Love the overview!

**Reply by rohitgarud** [reply]
Thank you for your answer to my query. I was able to use the config file to perform this simulation.
